spec.bs

```diff
@@ -5673,20 +5717,23 @@ of the following global objects:
       |global|.
     1. Return « |resultJSON|, |reportURL|,
       |global|'s [=InterestGroupReportingScriptRunnerGlobalScope/reporting beacon map=], |macroMap|,
-      |paContributions| ».
+      |paContributions|, |executionMetrics| ».
 </div>
 
 <div algorithm>
   To <dfn>evaluate a script</dfn> with a [=ECMAScript/realm=] |realm|, [=string=] |script|, [=string=]
   |functionName|, a [=list=] |arguments|, and an integer millisecond [=duration=] |timeout|, run these steps.
-  They return a [=ECMAScript/Completion Record=], which is either an [=ECMAScript/abrupt completion=] (in
-  the case of a parse failure or execution error), or a [=ECMAScript/normal completion=] populated with the
-  [=ECMAScript/ECMAScript language value=] result of invoking |functionName|.
+  They return a tuple of a [=ECMAScript/Completion Record=], which is either an [=ECMAScript/abrupt
+  completion=] (in the case of a parse failure or execution error), or a [=ECMAScript/normal
+  completion=] populated with the [=ECMAScript/ECMAScript language value=] result of invoking
+  |functionName|, and a [=boolean=] stating whether the script was interrupted due to reaching
+  |timeout|.
 
     1. [=Assert=] that these steps are running [=in parallel=].
 
-    1. If |timeout| &le; 0, [=immediately=] interrupt the execution and set |finalCompletion| to a
-      new [=ECMAScript/throw completion=] given null.
+    1. If |timeout| &le; 0, return (new [=ECMAScript/throw completion=] given null, true).

```

Comment:
.... good catch, though this needs more work: the timeout applies to top-level execution as well.  I am gonna need an extra moment on that.

(Actually it also affects WebIDL conversions of return values, since otherwise one could smuggle entire computation into them, but I am not sure how one even hooks that)

Reply:
Fixed, but kinda hand-wavish. PTAL.

Reply:
for webIDL conversion you mentioned, is it https://github.com/WICG/turtledove/blob/main/spec.bs#L5620?

Reply:
Yeah. That can run arbitrary code[1] (so it probably ought to be explicit about what context it runs in).
Same for scoreAd. Compare to:

https://source.chromium.org/chromium/chromium/src/+/main:content/services/auction_worklet/seller_worklet.cc;drc=b41d6ee12deac18fac89f4d27c2a662e311406fc;l=1232

(which I think I owe refactor of, but it's a little easier to read for this purpose than the bidder equivalent since the business stuff isn't hidden in the binding. Notice how it uses the same time limit; the v8_helper also provides the context).

[1] We have a couple other places this used where we start with more restricted types (sometimes infra types) where nothing funky can happen, as it's doing some sort of "round a double to int" type deals. It's where the source is an Object that JS gets... JS.

Reply:
Oh, and a testcase:
https://source.chromium.org/chromium/chromium/src/+/main:content/services/auction_worklet/bidder_worklet_unittest.cc;drc=33161dc8d856d514f774c8252763cc49bdc83c42;l=1539

Reply:
I think the context isn't explicit in the way the WebIDL and ES262 specs works, e.g. it calls ops like this:
https://tc39.es/ecma262/multipage/abstract-operations.html#sec-tonumber

Compare to:
https://source.chromium.org/chromium/chromium/src/+/main:v8/include/v8-value.h;drc=6ac7875f61c72fa111bc501039e4673693be6b64;l=384

...so I suppose returning the remaining timeout and having a similar thing about timeouts on those spec is appropriate.

Reply:
sounds good.